content/posts/2025-01-15-perf-improvements.md

```diff
@@ -0,0 +1,235 @@
+
+It's maybe less impressive, but is still _circa_ -20%, which can make a difference. In fact, it _does_ make a difference, because we have seen the overall cluster throughput slightly increase with these changes in our Test bed 1 (which is more stressed than Test bed 2 in terms of ingress traffic per node). To put it differently, NetObserv shows less overhead in stressed situations. More on that later.
+
+**Fig. 3: Total cluster traffic in Test bed 1**
+
+![Total traffic]({page.image('perf-improvements-1-8/total-traffic.png')})
+
+However, there's a catch: more traffic means... more flows to observe. In some ways, we get hit by our improvements. While we observed +18% traffic in the cluster, there's a parallel +11% memory increase in NetObserv because there is more to observe. It can be mitigated in multiple ways, such as with sampling, or by configuring the new filtering and conditional sampling options that we are also adding in this release.
+
+The performance enhancements also impact the kernel memory usage. As we'll see later, we did several changes on the BPF maps that optimize the memory used. We observed between -20% and -60% of the eBPF objects size in the hey-based tests (however, note that these tests don't focus on stressing memory).
+
+**Fig. 4: eBPF agent kernel-space memory usage, on a single node**
+
+![Kernel-space memory]({page.image('perf-improvements-1-8/kernel-space-memory.png')})
+
+_NB: the "wide load" profile used here corresponds to 3000 workers rate-limited at 2 qps, using a more distributed pattern across pods/nodes/namespaces._
+
+### Overhead
+
+Some resource overhead is inevitable when observing traffic, but hopefully NetObserv makes it as low as possible. By lowering the CPU usage, we expect to lower the overhead: less CPU used by the eBPF agent means more CPU available for other tasks. When running low on CPU, it can make a difference.
+
+I used _hey_ again to measure the overhead in two different ways:
+
+- in terms of added latency
+- in terms of maximum throughput
+
+After stress-testing a target, _hey_ provides statistics such as latency percentiles, min/max/average, and the number of queries per second. It can be configured with rate limits, but for the purpose of this test I didn't use it. It's only limited by the number of workers, which is set to 50. I am also running a rate-limited _hey-ho_ in parallel, just to generate some background noise and warm NetObserv.
+
+Here is the output of the baseline run, without NetObserv:
+
+```
+Summary:
+  Total:	30.0017 secs
+  Slowest:	0.1207 secs
+  Fastest:	0.0006 secs
+  Average:	0.0015 secs
+  Requests/sec:	39863.3677
+  
+  Total data:	735520320 bytes
+  Size/request:	735 bytes
+
+Response time histogram:
+  0.001 [1]	|
+  0.013 [999261]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
+  0.025 [538]	|
+  0.037 [70]	|
+  0.049 [32]	|
+  0.061 [38]	|
+  0.073 [9]	|
+  0.085 [1]	|
+  0.097 [0]	|
+  0.109 [0]	|
+  0.121 [50]	|
```

Comment:
Is each row the number of queries (HTTP requests?) in square brackets with the stated response time (first number)?

Is it because 0.013 is so dominant that the other values (even 538) is just a single "|"?

What does `0.0.97 [0]` and `0.109 [0]` mean?

Reply:
rows are the histogram buckets, and yes I think what's between brackets are the number of requests/responses in each bucket.
So there's indeed the vast majority of requests falling between 0.001 and 0.013 seconds

> Is it because 0.013 is so dominant that the other values (even 538) is just a single "|"?

yes it seems so

> What does 0.0.97 [0] and 0.109 [0] mean?

There's been 0 requests between 0.085 and 0.097 seconds, and also 0 between 0.097 and 0.109

Reply:
Okay, I was wondering why it even lists the "0 requests" unless these are preset time ranges.

Reply:
Well, the buckets are not always the same, so I believe they're adapted dynamically during the test, not sure exactly how.

Reply:
/lgtm